src/pass/storage_rewrite.cc

```diff
@@ -799,23 +819,17 @@ class StoragePlanRewriter : public IRMutator {
     if (const_nbits != 0) {
       // constant allocation.
       auto begin = const_free_map_.lower_bound(const_nbits / match_range);
-      auto mid = const_free_map_.lower_bound(const_nbits);
       auto end = const_free_map_.upper_bound(const_nbits * match_range);
-      for (auto it = mid; it != end; ++it) {
+      for (auto it = begin; it != end; ++it) {
         StorageEntry *e = it->second;
-        if (e->attach_scope_ != attach_scope) continue;
-        if (e->scope != scope) continue;
-        if (e->elem_type != op->type.element_of()) continue;
-        e->const_nbits = std::max(const_nbits, e->const_nbits);
-        const_free_map_.erase(it);
-        return e;
-      }
-      for (auto it = mid; it != begin;) {
-        --it;
-        StorageEntry *e = it->second;
-        if (e->attach_scope_ != attach_scope) continue;
-        if (e->scope != scope) continue;
-        if (e->elem_type != op->type.element_of()) continue;
+        if (!CanReuse(e, op, attach_scope, scope)) continue;
+        // when space left, try reuse it the next var, eg. two int16 after a float32
+        if (const_nbits < e->const_nbits) {
+          StorageEntry *new_e = NewAlloc(op, attach_scope, scope, e->const_nbits - const_nbits);
+          new_e->bits_offset = e->bits_offset + const_nbits;
```

Comment:
It is not correct to update bit_offset of the buffer since we are doing reuse(instead of merging)

Reply:
reuse means the two buffer can share same memory without conflicting lifecycle. Merging means two buffer cannot share the same memory region.

Reply:
I'll call it buffer split. add for this case: C and D can reuse A1 with this patch
```
with ib.for_range(0, 10, name="j") as j:
 +            A = ib.allocate("float32", 200, name="A", scope="local.L0A")
 +            A1 = ib.allocate("float32", 200, name="A1", scope="local.L0A")
 +            A[j] = 1.2
 +            A1[j] = 1.3
 +            B = ib.allocate("int16", 200, name="B", scope="local.L0A")
 +            B[j] = tvm.const(1, "int16")
 +            C = ib.allocate("int16", 200, name="C", scope="local.L0A")
 +            C[j] = tvm.const(1, "int16")
 +            D = ib.allocate("int16", 200, name="D", scope="local.L0A")
 +            D[j] = B[j] + C[j]
 +            A2 = ib.allocate("float32", 200, name="A2", scope="local.L0A")
 +            A2[j] = A[j]
```

Reply:
I am not so sure about striping buffers into two, and whether current logic correctly handles this. Maybe remove this logic and propose the more robust version again in an issue or next PR

Reply:
C(100 in float32) reuses the whole A1(200 in float32) with current logic.

Reply:
another PR is ok with me